### PR TITLE
p2p: Redesign the fast-sync mechanism

### PR DESCRIPTION
**Pull request by gsalgado**
We now fetch a batch of headers and then all the bodies/receipts for those
headers before persisting them. This makes the code a lot easier to follow
but also guarantees our ChainDB always has full blocks for everything up to
the current chain head.

### CODE REVIEW

File: p2p/chain.py [hunk 1]
```diff
@@ -164,142 +149,134 @@ def register_peer(self, peer: BasePeer) -> None:
...
+            missing = [
+                header for header in missing
+                if key_func(header) not in downloaded]
+
+    def _request_block_parts(
+            self,
+            headers: List[BlockHeader],
+            request_func: Callable[[ETHPeer, List[BlockHeader]], None]) -> int:
+        length = math.ceil(len(headers) / len(self.peer_pool.peers))
+        batches = list(partition_all(length, headers))
```
Comment (pipermerriam) [inline p2p/chain.py]: I assume the `list` casting here has to do with the following `zip` not supporting whatever `partition_all` natively returns?

File: p2p/chain.py [hunk 2]
```diff
@@ -164,142 +149,134 @@ def register_peer(self, peer: BasePeer) -> None:
...
-                peer
-                for peer in self.peer_pool.peers
-                if peer not in self._peers_with_pending_requests
-            ]
-            if idle_peers:
-                return cast(ETHPeer, secrets.choice(idle_peers))
-            else:
-                self.logger.debug("No idle peers availabe, sleeping a bit")
-                await asyncio.sleep(0.2)
+    def _body_key(self, header: BlockHeader) -> Tuple[bytes, bytes]:
```
Comment (pipermerriam) [inline p2p/chain.py]: Maybe elevate these to top level functions to get them out of the class namespace?

File: p2p/chain.py [hunk 3]
```diff
@@ -164,142 +149,134 @@ def register_peer(self, peer: BasePeer) -> None:
...
+        """Return the unique key of the body for the given header.
+
+        i.e. a two-tuple with the transaction root and uncles hash.
+        """
+        return cast(Tuple[bytes, bytes], (header.transaction_root, header.uncles_hash))
+
+    def _is_body_empty(self, header: BlockHeader) -> bool:
```
Comment (pipermerriam) [inline p2p/chain.py]: Same (elevate to top level module function)

File: p2p/chain.py [hunk 4]
```diff
@@ -164,142 +149,134 @@ def register_peer(self, peer: BasePeer) -> None:
...
+        return False
+
+    def request_bodies(self, headers: List[BlockHeader]) -> int:
+        """Ask our peers for bodies for the given headers.
+
+        See request_receipts() for details of how this is done.
+        """
+        return self._request_block_parts(headers, self._send_get_block_bodies)
+
+    def _receipts_key(self, header: BlockHeader) -> bytes:
```
Comment (pipermerriam) [inline p2p/chain.py]: And so on about elevating to top level module function, not going to comment on the others.

File: p2p/chain.py [hunk 5]
```diff
@@ -378,6 +325,35 @@ def _skip_empty_and_duplicated_receipts(
...
+        self.logger.debug("Got Receipts for %d blocks from %s", len(receipts), peer)
+        loop = asyncio.get_event_loop()
+        iterator = map(make_trie_root_and_nodes, receipts)
+        receipts_tries = await wait_with_token(
+            loop.run_in_executor(None, list, iterator),
+            token=self.cancel_token)
+        receipt_roots = []
+        for receipt_root, trie_dict_data in receipts_tries:
+            await self.chaindb.coro_persist_trie_data_dict(trie_dict_data)
+            receipt_roots.append(receipt_root)
```
Comment (pipermerriam) [inline p2p/chain.py]: What do you think about changing this to the following (after the exit of the loop)

```python
receipt_roots = tuple(zip(*receipt_tries))[0]
```

File: p2p/chain.py [hunk 6]
```diff
@@ -378,6 +325,35 @@ def _skip_empty_and_duplicated_receipts(
...
+
+    async def _handle_block_bodies(self, peer: ETHPeer, bodies: List[eth.BlockBody]) -> None:
+        self.logger.debug("Got Bodies for %d blocks from %s", len(bodies), peer)
+        loop = asyncio.get_event_loop()
+        iterator = map(make_trie_root_and_nodes, [body.transactions for body in bodies])
+        transactions_tries = await wait_with_token(
+            loop.run_in_executor(None, list, iterator),
+            token=self.cancel_token)
+        body_keys = []
+        for i, body in enumerate(bodies):
```
Comment (pipermerriam) [inline p2p/chain.py]: I think you can change this to be:

```python
for (body, (tx_root, trie_data_dict)) in zip(bodies, transaction_tries):
    ...
```

File: p2p/chain.py (hunk 1 above)
Comment (gsalgado) [reply p2p/chain.py]: Yep, otherwise we can't get its length

File: p2p/chain.py (hunk 3 above)
Comment (gsalgado) [reply p2p/chain.py]: In order to more this one I'd need to add an extra argument (for the empty_root_hash, which it currently takes from the chaindb), so I'd rather keep it here unless you feel strongly about it

Comment (pipermerriam) [reply p2p/chain.py]: You can actually hard code that value as the alternate trie implementation which required the empty root hash to be specified is being removed as part of #518

File: p2p/chain.py (hunk 4 above)
Comment (gsalgado) [reply p2p/chain.py]: Moved as well

File: p2p/chain.py (hunk 5 above)
Comment (gsalgado) [reply p2p/chain.py]: I'd prefer the current version as it doesn't have any cognitive overhead and when debugging one can immediately see the list will contain all receipt roots (and just them) at the end of the loop, whereas the latter requires me to remember/lookup how zip works, and in fact I wasn't even aware it is a variadic function

File: p2p/chain.py (hunk 6 above)
Comment (gsalgado) [reply p2p/chain.py]: Indeed!

File: p2p/chain.py (hunk 5 above)
Comment (pipermerriam) [reply p2p/chain.py]: Ok, how about:

```python
receipt_roots = [receipt_root for receipt_root, _ in receipt_tries]
```

Simply removes the loop based accumulation in favor of a comprehension which maintains readability.

Comment (pipermerriam) [reply p2p/chain.py]: Just an FYI so you know it exists, I just found [`cytoolz.pluck`](http://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.pluck) 

```python
receipt_roots = tuple(pluck(0, receipt_tries))  # cast to tuple to force evaluation
```

**Pull request merged**